### ISSUE TITLE
Type Guad Narrowing using 'in' operator not working

### ISSUE THREAD

**Issue by TheColorRed** [Microsoft/TypeScript#20893]
**TypeScript Version:**  2.6.2
**TypeScript Next Version:**  2.7.0-dev.20171223

When using `if('property' in object)` type guard doesn't seem to pick it up. 
For example: `if ('length' in obj && obj.length == 0)` the error says that length can not be found:

> Property 'length' does not exist on type

Sometimes using `instanceof` doesn't work due to the fact that you may get a `circular reference`, so the code then just breaks. Using `in` and not getting an error would be helpful.

**Code**

**OtherClass.ts**
If I were to switch the `in` with `instanceof` here, I would get a `circular reference`
```ts
// Simple example class that does nothing
class OtherClass {
  public test() {
    let obj = this.other() // Returns an object that may or many not have `length`
    if ('length' in obj && obj.length == 0) {
      throw new Error('Oops array length too short')
    }
  }
}
```
**MyClass.ts**
```ts
import {OtherClass} from './OtherClass'
// Simple example class with an array and a length getter
class MyClass extends OtherClass {
  private items: {}[] = []
  public get length() { return this.items.length }
}
```
**Usage:**
```ts
// Get one of the classes at random
let obj = new MyClass()
obj.test()
```

**Expected behavior:**
`obj.length` shouldn't give an error

**Actual behavior:**
`obj.length` says:
> Property 'length' does not exist on type

![javascript in](https://user-images.githubusercontent.com/1571806/34342071-b277262c-e96a-11e7-94d9-9a6f0a10a27f.png)

**Comment by sylvanaar** [comment]
You should really provide the type of `this.other()` to narrow down you issue.

Based on the error message though, it seems you have a union type, and I guess only one of the types in the union has the length property.

If you think about how you would accomplish what you are trying to do in javascript it may help you arrive at a solution.

How about the following:

`
type Foo = { foo: number }
type Bar = { bar: number, length: number }
type Baz = Foo | Bar

class OtherClass {
    public test() {
      let obj: Baz // Returns an object that may or many not have `length`
      if ((<Bar>obj).length === 0) {
        throw new Error('Oops array length too short')
      }
    }
  }`

**Comment by sylvanaar** [comment]
Sorry for earlier comment, seems like the `in` operator type narrowing is a 2.7 feature.

I tested on `2.7.0-dev.20171226`, and observed the correct "expected" behavior.

**Comment by TheColorRed** [comment]
This seems to have been fixed.
